### ISSUE TITLE
[Suggestion] Shared Experience

### ISSUE THREAD

**Issue by edumntg** [otland/forgottenserver#1343]
Move shared-experience mechanism to lua (Player events?)
##

**Comment by Kamenuvol** [comment]
If I had time, I would have started working on it. My idea is to create a function to set the criteria of dividing xp or set a Party Rate on a new Party:onShareExperience event.
And on Player:onGainExperience multiply the rate returned by the function, or an actual party rate.

**Comment by Kamenuvol** [comment]
Also the exp sharing thing is completely buggy. The experience values showed on the server log and on the game screen are different and the formula is wrong.

**Comment by ranisalt** [comment]
@Kamenuvol could you explain further on why the formula is wrong? We had some discussion here some time ago and concluded the formula is correct and only misses by a few points, related to flooring or ceiling the result.

**Comment by Kamenuvol** [comment]
```c++
uint32_t shareExperience = static_cast<uint64_t>(std::ceil(((static_cast<double>(experience) / (memberList.size() + 1)) + (static_cast<double>(experience) * extraExpRate))));
```
As we see, we are doing (exp)*(1/size)  + (exp)*(rate) or (exp)*(rate + 1/size)
The correct formula is: (exp)*(rate)/size

e.g: A party with 4 players with 3 different vocations kills a monster that gives 800xp.
Current formula: 800*(0.6 + 0.25) = 800*0.85 = 680
Correct: 800*1.6*0.25 = 320

Source: http://tibia.wikia.com/wiki/Party

**Comment by edumntg** [comment]
Correct formula is ((exp*party_bonus)/size)*player_bonus

party_bonus = shared experience bonus (+20% for sharing +x% depending of vocations in the party)
size = size of party
player_bonus = green stamina?

e.g: A knight inside a party with 2 members (paladin and druid), and the knight has green stamina. The kill a demon:

exp = 6000
party_bonus = +20% +40%(3 different vocations) = +60% = 1.6
size = 3
player_bonus = +50% = 1.5

Total experience = ((6000*1.8)/3)*1.5 = 4800

**Comment by Kamenuvol** [comment]
I've omitted the player bonus because it is already in Player events. Party shared exp is indepedent.

**Comment by ranisalt** [comment]
Please have a look at the PR.

**Comment by sircurse** [comment]
Hi folks, I managed to add the chagnes that Renisalt proposed, and it worked great for me, it must be a fix not a enhancement.

**Comment by vankk** [comment]
@DSpeichert You forgot to close this issue since it was approved in #2478

**Comment by DSpeichert** [comment]
Fixed in #2478
